src/frontend/language_phase.nit

```diff
@@ -0,0 +1,121 @@
+# Basic support of internationalization through the generation of id-to-string tables
+module language_phase
+
+private import annotation
+private import parser_util
+import astbuilder
+import gettext
+
+redef class ToolContext
+	# Main phase of `language`
+	var language_phase: Phase = new LanguagePhase(self, null)
+end
+
+class LanguagePhase
+	super Phase
+
+	redef fun process_annotated_node(nmodule, nat) do
+		if not nat.name == "language" then return
+
+		if not nmodule isa AModuledecl then
+			toolcontext.error(nmodule.location, "Cannot use annotation language on something other than a module")
+			return
+		end
+
+		var domain = nmodule.n_name.n_id.text
+
+		if nat.n_args.length != 1 then
+			toolcontext.error(nat.location, "Error: wrong number of arguments for annotation language, expected 1, got {nat.n_args.length}")
+			return
+		end
+		var arg = nat.n_args[0]
+
+		if not arg isa AStringFormExpr then
+			toolcontext.error(nat.location, "Error: expected string argument for annotation language")
+			return
+		end
+
+		var lang_raw = arg.n_string.text
+		var lang = lang_raw.substring(1, lang_raw.length - 2)
+
+		if lang == "" then
+			toolcontext.error(nat.location, "Error: invalid value for argument of annotation, expecting non-empty string")
+		end
+
+		var foldr = nmodule.location.to_s.split("/")
+		foldr.pop
+		var foldr_path = foldr.join("/").realpath
+		foldr_path += "/languages"
+
+		if not foldr_path.file_exists then foldr_path.mkdir
+
+		var lng_foldr = foldr_path + "/{lang}"
+
+		if not lng_foldr.file_exists then lng_foldr.mkdir
+
+		var mod = nmodule.parent.as(AModule)
+
+		var vi = new StringFinder(domain, foldr_path, toolcontext)
+		vi.enter_visit(mod)
+
+		var modl = nmodule.location.file.filename.basename(".nit")
+
+		var fnm = foldr_path + "/" + modl
+		var arr = new Array[PObject]
+		for v in vi.strings.values do arr.push v
+		var po = new POFile.with_strings(arr)
+		po.write_template(fnm)
+		for i in po.strings do
+			i.msgstr = i.msgid
+		end
+
+		var msg_foldr = lng_foldr + "/LC_MESSAGES"
+		if not msg_foldr.file_exists then msg_foldr.mkdir
+
+		po.write_to_file(msg_foldr + "/" + modl)
+	end
+end
+
+private class StringFinder
+	super Visitor
+
+	var strings = new HashMap[String, PObject]
+
+	# Domain of the strings to internationalize
+	var domain: String
+	# Location of the languages file
+	var languages_location: String
+
+	var toolcontext: ToolContext
+
+	redef fun visit(n)
+	do
+		n.accept_string_finder(self)
+		n.visit_all(self)
+	end
+end
+
+redef class ANode
+	private fun accept_string_finder(v: StringFinder) do end
+end
+
+redef class AStringFormExpr
+
+	redef fun accept_string_finder(v) do
```

Comment:
Yes, but since gettext takes the key if there's not a msgid assigned to it in the target language, I figured it would not be too much of a hassle ?